### PR TITLE
Adds ``--base-path` to solc when compiling in `--standard-json` mode for resolving relative paths.

### LINKED ISSUE #5280: "allow-paths" is not used correctly for imports in standard json
EDIT:

Add option `--base-path` for standard-json CLI such that the read callback reads relative to that.

---

## Description

`--allow-paths` does not work correctly when importing files with standard json. All imports, even relative to the current file are disallowed. 

When using command line flags, instead of `--standard-json` it works just fine.

## Environment

- Compiler version:
   * 0.5.0-develop.2018.10.12+commit.94526b2d.Linux.g++
   * 0.4.25+commit.59dbf8f1.Linux.g++
- Operating system:
   * Linux (Ubuntu 18)

## Steps to Reproduce

Create the following structure in your home directory:
```
bug/
    bug.json
    file.sol
    test.sol
```

file.sol
```
pragma solidity ^0.4.25;
import "./test.sol";
```

test.sol
```
pragma solidity ^0.4.25;
contract Token {
}
```

bug.json
```
{
  "language": "Solidity",
  "sources": {
    "file.sol": {
      "urls": [
        "/home/user/bug/file.sol"
      ]
    }
  },
  "settings": {
    "evmVersion": "byzantium",
    "outputSelection": {
      "*": {
        "*": [
          "abi"
        ]
      }
    }
  }
}
```

Now navigate to the directory above and run the following:
```
cat bug/bug.json | solc --allow-paths /home/user/bug --standard-json
```
The compiler will find `file.sol` just fine, but will complain that `test.sol` is not in an allowed directory.

### PR DESCRIPTION
**Pull request by christianparpart**
Closes https://github.com/ethereum/solidity/issues/5280

### CODE REVIEW

File: solc/CommandLineInterface.cpp [hunk 1]
```diff
@@ -828,6 +830,11 @@ Allowed options)").c_str(),
 			po::value<string>()->value_name("path(s)"),
 			"Allow a given path for imports. A list of paths can be supplied by separating them with a comma."
 		)
+		(
+			g_argBasePath.c_str(),
+			po::value<string>()->value_name("path"),
+			"Allow the given path to be used as base path for relative imports."
```
Comment (chriseth) [inline solc/CommandLineInterface.cpp]: Actually ti si for absolute imports :)

Comment (chriseth) [reply solc/CommandLineInterface.cpp]: Use the given path as the root of the source tree instead of the root of the filesystem. Implies allowing access to that path for imports.

File: solc/CommandLineInterface.cpp [hunk 2]
```diff
@@ -963,6 +970,12 @@ bool CommandLineInterface::processInput()
 			string validPath = _path;
 			if (validPath.find("file://") == 0)
 				validPath.erase(0, 7);
+			else if (validPath[0] == '.' ||
+					(validPath[0] == '/' && !boost::filesystem::exists(validPath)) ||
+					validPath[0] != '/')
+			{
+				validPath = m_basePath.string() + "/" + validPath;
+			}
```
Comment (chriseth) [inline solc/CommandLineInterface.cpp]:
```suggestion
			auto path = boost::filesystem::path(m_basePath.string()) / validPath;
```

(if that works for the empty string)

File: solc/CommandLineInterface.cpp [hunk 3]
```diff
@@ -1001,6 +1014,20 @@ bool CommandLineInterface::processInput()
...
+		auto const fspath = boost::filesystem::path(m_args[g_argBasePath].as<string>());;
+		if (!boost::filesystem::is_directory(fspath))
+		{
+			serr() << "Base path must be a directory. \"" << fspath << "\"\n";
+			return false;
+		}
+		else
+		{
+			m_basePath = fspath;
+		}
```
Comment (chriseth) [inline solc/CommandLineInterface.cpp]:
```suggestion
		m_basePath = fspath;
```

File: solc/CommandLineInterface.cpp [hunk 4]
```diff
@@ -1001,6 +1014,20 @@ bool CommandLineInterface::processInput()
 		}
 	};
 
+	if (m_args.count(g_argBasePath))
+	{
+		auto const fspath = boost::filesystem::path(m_args[g_argBasePath].as<string>());;
```
Comment (chriseth) [inline solc/CommandLineInterface.cpp]:
```suggestion
		boost::filesystem::path const fspath{m_args[g_argBasePath].as<string>()};
```

**Review by chriseth — CHANGES_REQUESTED** [review]
.

File: Changelog.md [hunk 5]
```diff
@@ -28,6 +28,7 @@ Compiler Features:
  * Commandline Interface: Don't ignore `--yul-optimizations` in assembly mode.
  * Allow using abi encoding functions for calldata array slices without explicit casts.
  * Wasm binary output: Implement ``br`` and ``br_if``.
+ * Commandline Interface: Adds new option ``--base-path PATH`` for resolving relative paths when importing files via ``--standard-json``
```
Comment (chriseth) [inline Changelog.md]: Needs to be moved.

File: solc/CommandLineInterface.cpp [hunk 6]
```diff
@@ -963,7 +970,8 @@ bool CommandLineInterface::processInput()
 			string validPath = _path;
 			if (validPath.find("file://") == 0)
 				validPath.erase(0, 7);
-			auto path = boost::filesystem::path(validPath);
+
+			auto const path = boost::filesystem::path(m_basePath.string()) / validPath;
```
Comment (chriseth) [inline solc/CommandLineInterface.cpp]: Why convert to string and back to path?

File: solc/CommandLineInterface.cpp [hunk 7]
```diff
@@ -1001,6 +1009,17 @@ bool CommandLineInterface::processInput()
 		}
 	};
 
+	if (m_args.count(g_argBasePath))
+	{
+		boost::filesystem::path const fspath{m_args[g_argBasePath].as<string>()};
+		if (!boost::filesystem::is_directory(fspath))
+		{
+			serr() << "Base path must be a directory. \"" << fspath << "\"\n";
```
Comment (chriseth) [inline solc/CommandLineInterface.cpp]:
```suggestion
			serr() << "Base path must be a directory: \"" << fspath << "\"\n";
```

**Comment by chriseth** [comment]
We should also explain that in "using the compiler" in the documentation.

**Comment by chriseth** [comment]
We also need to improve the documenation around this area - see https://github.com/ethereum/solidity/issues/5280#issuecomment-628268525

File: solc/CommandLineInterface.cpp (hunk 6 above)
Comment (christianparpart) [reply solc/CommandLineInterface.cpp]: you are right.

File: Changelog.md [hunk 8]
```diff
@@ -28,6 +28,7 @@ Language Features:
 
 Compiler Features:
  * Commandline Interface: Don't ignore `--yul-optimizations` in assembly mode.
+ * Commandline Interface: Adds new option ``--base-path PATH`` for resolving relative paths when importing files via ``--standard-json``
```
Comment (chriseth) [inline Changelog.md]: needs to move

File: docs/using-the-compiler.rst [hunk 9]
```diff
@@ -44,6 +44,10 @@ An empty remapping prefix is not allowed.
 
 If there are multiple matches due to remappings, the one with the longest common prefix is selected.
 
+When importing files, all to be imported files are treated as if they were fully qualified paths,
+unless remapped as earlier shown. This behaviour can be customized by adding the command line
+option ``--base-path`` with a path to be prepended before each import is performed.
+
```
Comment (chriseth) [inline docs/using-the-compiler.rst]:
```suggestion
When accessing the filesystem to search for imports, all paths are treated as if they were fully qualified paths.
This behaviour can be customized by adding the command line
option ``--base-path`` with a path to be prepended before each filesystem access for imports is performed.
```

File: docs/using-the-compiler.rst [hunk 10]
```diff
@@ -44,6 +44,10 @@ An empty remapping prefix is not allowed.
...
 For security reasons the compiler has restrictions what directories it can access. Paths (and their subdirectories) of source files specified on the commandline and paths defined by remappings are allowed for import statements, but everything else is rejected. Additional paths (and their subdirectories) can be allowed via the ``--allow-paths /sample/path,/another/sample/path`` switch.
```
Comment (chriseth) [inline docs/using-the-compiler.rst:55]: Everything inside the path specified via `--base-path` is always allowed.

File: Changelog.md (hunk 8 above)
Comment (christianparpart) [reply Changelog.md]: okay, I was blind. wrong version section.

File: solc/CommandLineInterface.cpp [hunk 11]
```diff
@@ -1001,6 +1009,17 @@ bool CommandLineInterface::processInput()
...
 
+	if (m_args.count(g_argBasePath))
+	{
+		boost::filesystem::path const fspath{m_args[g_argBasePath].as<string>()};
+		if (!boost::filesystem::is_directory(fspath))
+		{
+			serr() << "Base path must be a directory; \"" << fspath << "\"\n";
+			return false;
+		}
+		m_basePath = fspath;
```
Comment (chriseth) [inline solc/CommandLineInterface.cpp:1022]: Please add it to the allowed paths.

File: docs/using-the-compiler.rst [hunk 12]
```diff
@@ -44,7 +44,13 @@ An empty remapping prefix is not allowed.
 
 If there are multiple matches due to remappings, the one with the longest common prefix is selected.
 
-For security reasons the compiler has restrictions what directories it can access. Paths (and their subdirectories) of source files specified on the commandline and paths defined by remappings are allowed for import statements, but everything else is rejected. Additional paths (and their subdirectories) can be allowed via the ``--allow-paths /sample/path,/another/sample/path`` switch.
+When accessing the filesystem to search for imports, all paths are treated as if they were fully qualified paths.
+This behaviour can be customized by adding the command line option ``--base-path`` with a path to be prepended before each filesystem access for imports is performed.
+For security reasons the compiler has restrictions what directories it can access.
```
Comment (chriseth) [inline docs/using-the-compiler.rst]: new paragraph

File: solc/CommandLineInterface.cpp [hunk 13]
```diff
@@ -1001,6 +1010,19 @@ bool CommandLineInterface::processInput()
...
+		if (!boost::range::count(m_allowedDirectories, fspath))
```
Comment (chriseth) [inline solc/CommandLineInterface.cpp]:
```suggestion
		if (!contains(m_allowedDirectories, fspath))
```

File: Changelog.md [hunk 14]
```diff
@@ -5,6 +5,7 @@ Language Features:
 
 Compiler Features:
  * Build system: Update the soljson.js build to emscripten 1.39.15 and boost 1.73.0 and include Z3 for integrated SMTChecker support without the callback mechanism.
+ * Commandline Interface: Adds new option ``--base-path PATH`` for resolving relative paths when importing files via ``--standard-json``
```
Comment (chriseth) [inline Changelog.md]: Thus should work both in standard-json as well as in the regular commandline interface, should it?

Also,  maybe rather use the wording from the help screen? This is not necessarily about relative paths.

**Comment by chriseth** [comment]
Will take a look here to get this merged.

File: docs/using-the-compiler.rst [hunk 15]
```diff
@@ -44,7 +44,16 @@ An empty remapping prefix is not allowed.
 
 If there are multiple matches due to remappings, the one with the longest common prefix is selected.
 
-For security reasons the compiler has restrictions what directories it can access. Paths (and their subdirectories) of source files specified on the commandline and paths defined by remappings are allowed for import statements, but everything else is rejected. Additional paths (and their subdirectories) can be allowed via the ``--allow-paths /sample/path,/another/sample/path`` switch.
+When accessing the filesystem to search for imports, all paths are treated as if they were fully qualified paths.
+This behaviour can be customized by adding the command line option ``--base-path`` with a path to be prepended
+before each filesystem access for imports is performed. Furthermore, the part added via ``--base-path``
+will not appear in the contract metadata.
+
+For security reasons the compiler has restrictions what directories it can access.
```
Comment (hrkrshnn) [inline docs/using-the-compiler.rst]:
```suggestion
For security reasons the compiler has restrictions on what directories it can access.
```

File: solc/CommandLineInterface.cpp [hunk 16]
```diff
@@ -1003,6 +1011,19 @@ bool CommandLineInterface::processInput()
 		}
 	};
 
+	if (m_args.count(g_argBasePath))
+	{
+		boost::filesystem::path const fspath{m_args[g_argBasePath].as<string>()};
+		if (!boost::filesystem::is_directory(fspath))
+		{
+			serr() << "Base path must be a directory; \"" << fspath << "\"\n";
```
Comment (leonardoalt) [inline solc/CommandLineInterface.cpp]:
```suggestion
			serr() << "Base path must be a directory: \"" << fspath << "\"\n";
```
?

**Review by leonardoalt — COMMENTED** [review]
This is probably good for now, but I'm wondering whether it makes sense to allow multiple `--bash-path`s, like `-I` in GCC.

**Comment by chriseth** [comment]
Could you include my changes from https://github.com/ethereum/solidity/commit/2d0089182584b07cb71299f0ca2e9d474dd5413a please?

File: docs/standard-json-format.rst [hunk 17]
```diff
@@ -0,0 +1,112 @@
+.. index:: standard.json file format
```
Comment (chriseth) [inline docs/standard-json-format.rst]: Can we do this in its own PR?

Comment (christianparpart) [reply docs/standard-json-format.rst]: that's what i initially wanted. I'll move it out, sure.

**Review by chriseth — APPROVED** [review]
